### ISSUE TITLE
MPI.Gatherv! should only require counts for root

### ISSUE THREAD

**Issue by mfsch** [JuliaParallel/MPI.jl#362]
I was trying to use the following (simplified) code to gather data to the root process, where the size of the data can vary per process:

```
count_local = Cint(length(data))
counts_all = MPI.Gather(count_local, 0, MPI.COMM_WORLD)
data_all = MPI.Gatherv(data, counts_all, 0, MPI.COMM_WORLD)
```

However, this fails since `counts_all` is `nothing` for the non-root processes. and `Gatherv` requires the argument to be of type `Vector{Cint}`.

AFAIK this is a pretty common MPI usage pattern so it might make sense to support it. From looking at `Gatherv!` in `collective.jl`, it seems that `counts` is used to set `displ` and `sendcnt` at non-root processes. According to the [MPICH documentation](https://www.mpich.org/static/docs/v3.1/www3/MPI_Gatherv.html), `recvcounts` and `displs` should only be required for the root, and `sendcnt` could be set to `length(sendbuf)` as is done in some other collective functions. This way, the function signature could be changed to accept `Union{Vector{Cint},Nothing}` for the counts.

Is this something that you would like to support in `MPI.jl`? Or are there any drawbacks to this that I haven’t considered? If there’s interest, I should be able to prepare a pull request for this.

**Comment by simonbyrne** [comment]
Yes, this is something I would like to change. This should be possible with #335, but I don't know when I will have time to finish that up (though now I know there is demand, I may try to spend some time on it). In the meantime, if there is an easy way to fix it, I would be open to a PR.

**Comment by simonbyrne** [comment]
Fixed by #335.
